File: mobile-apps/android/KeylessEntry.kt
/**
 * KeylessEntry.kt
 * Fleet Manager - Keyless Vehicle Entry for Android
 *
 * Bluetooth Low Energy and NFC-based keyless entry system
 * with encrypted keys and time-limited access
 */

package com.fleet.manager.keyless

import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.nfc.*
import android.nfc.tech.Ndef
import android.os.ParcelUuid
import androidx.compose.runtime.*
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.security.KeyStore
import java.util.*
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import kotlin.math.pow

// MARK: - Keyless Entry Manager

class KeylessEntryManager(private val context: Context) {
    private val bluetoothAdapter: BluetoothAdapter? = BluetoothAdapter.getDefaultAdapter()
    private val bluetoothLeScanner: BluetoothLeScanner? = bluetoothAdapter?.bluetoothLeScanner
    private var bluetoothGatt: BluetoothGatt? = null
    private var nfcAdapter: NfcAdapter? = NfcAdapter.getDefaultAdapter(context)

    // Service UUIDs
    private val vehicleServiceUUID = UUID.fromString("12345678-1234-1234-1234-123456789ABC")
    private val unlockCharacteristicUUID = UUID.fromString("12345678-1234-1234-1234-123456789ABD")
    private val statusCharacteristicUUID = UUID.fromString("12345678-1234-1234-1234-123456789ABE")

    // State
    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning

    private val _nearbyVehicles = MutableStateFlow<List<VehicleBeacon>>(emptyList())
    val nearbyVehicles: StateFlow<List<VehicleBeacon>> = _nearbyVehicles

    private val _connectionStatus = MutableStateFlow(ConnectionStatus.DISCONNECTED)
    val connectionStatus: StateFlow<ConnectionStatus> = _connectionStatus

    private val _unlockStatus = MutableStateFlow(UnlockStatus.LOCKED)
    val unlockStatus: StateFlow<UnlockStatus> = _unlockStatus

    // Security
    private var vehicleKey: SecretKey? = null
    private var accessToken: String? = null
    private var tokenExpiry: Long? = null

    companion object {
        private const val KEYSTORE_ALIAS = "fleet_vehicle_key"
        private const val PREFS_NAME = "keyless_entry"
        private const val PREF_ACCESS_TOKEN = "access_token"
        private const val PREF_TOKEN_EXPIRY = "token_expiry"
    }

    init {
        loadStoredCredentials()
    }

    // MARK: - Credential Management

    private fun loadStoredCredentials() {
        // Load key from Android KeyStore
        try {
            val keyStore = KeyStore.getInstance("AndroidKeyStore")
            keyStore.load(null)

            if (keyStore.containsAlias(KEYSTORE_ALIAS)) {
                vehicleKey = keyStore.getKey(KEYSTORE_ALIAS, null) as? SecretKey
            }
        } catch (e: Exception) {
            println("Error loading key: ${e.message}")
        }

        // Load token and expiry from SharedPreferences
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        accessToken = prefs.getString(PREF_ACCESS_TOKEN, null)
        tokenExpiry = prefs.getLong(PREF_TOKEN_EXPIRY, 0).takeIf { it > 0 }
    }

    fun setVehicleCredentials(vehicleId: String, token: String, expiryHours: Int = 24) {
        // Generate and store encryption key
        val keyGenerator = KeyGenerator.getInstance("AES", "AndroidKeyStore")
        keyGenerator.init(256)
        vehicleKey = keyGenerator.generateKey()

        // Store access token
        accessToken = token
        tokenExpiry = System.currentTimeMillis() + (expiryHours * 3600 * 1000)

        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit().apply {
            putString(PREF_ACCESS_TOKEN, token)
            putLong(PREF_TOKEN_EXPIRY, tokenExpiry!!)
            apply()
        }

        println("Vehicle credentials set for $vehicleId, expires at ${Date(tokenExpiry!!)}")
    }

    private fun isTokenValid(): Boolean {
        val expiry = tokenExpiry ?: return false
        return System.currentTimeMillis() < expiry
    }

    // MARK: - Bluetooth Scanning

    fun startScanning() {
        if (!isTokenValid()) {
            println("Access token expired")
            return
        }

        _isScanning.value = true
        _nearbyVehicles.value = emptyList()

        val scanSettings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        val scanFilter = ScanFilter.Builder()
            .setServiceUuid(ParcelUuid(vehicleServiceUUID))
            .build()

        bluetoothLeScanner?.startScan(listOf(scanFilter), scanSettings, scanCallback)
        println("Started scanning for vehicles...")
    }

    fun stopScanning() {
        bluetoothLeScanner?.stopScan(scanCallback)
        _isScanning.value = false
        println("Stopped scanning")
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val rssi = result.rssi
            val name = device.name ?: "Unknown Vehicle"
            val distance = calculateDistance(rssi.toDouble())

            val vehicle = VehicleBeacon(
                id = device.address,
                name = name,
                distance = distance,
                rssi = rssi,
                device = device
            )

            val currentVehicles = _nearbyVehicles.value.toMutableList()
            if (!currentVehicles.any { it.id == vehicle.id }) {
                currentVehicles.add(vehicle)
                _nearbyVehicles.value = currentVehicles
                println("Discovered vehicle: $name at ${distance}m")
            }
        }

        override fun onScanFailed(errorCode: Int) {
            println("Scan failed with error: $errorCode")
            _isScanning.value = false
        }
    }

    // MARK: - Vehicle Connection

    fun connect(vehicle: VehicleBeacon) {
        _connectionStatus.value = ConnectionStatus.CONNECTING
        bluetoothGatt = vehicle.device?.connectGatt(context, false, gattCallback)
        println("Connecting to ${vehicle.name}...")
    }

    fun disconnect() {
        bluetoothGatt?.disconnect()
        bluetoothGatt?.close()
        bluetoothGatt = null
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    println("Connected to vehicle")
                    _connectionStatus.value = ConnectionStatus.CONNECTED
                    gatt.discoverServices()
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    println("Disconnected from vehicle")
                    _connectionStatus.value = ConnectionStatus.DISCONNECTED
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                val service = gatt.getService(vehicleServiceUUID)
                val statusChar = service?.getCharacteristic(statusCharacteristicUUID)

                statusChar?.let {
                    gatt.readCharacteristic(it)
                    gatt.setCharacteristicNotification(it, true)
                }
            }
        }

        override fun onCharacteristicRead(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (characteristic.uuid == statusCharacteristicUUID) {
                val status = characteristic.getStringValue(0)
                _unlockStatus.value = if (status == "unlocked") UnlockStatus.UNLOCKED else UnlockStatus.LOCKED
            }
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            if (characteristic.uuid == statusCharacteristicUUID) {
                val status = characteristic.getStringValue(0)
                _unlockStatus.value = if (status == "unlocked") UnlockStatus.UNLOCKED else UnlockStatus.LOCKED
            }
        }
    }

    // MARK: - Unlock/Lock Operations

    fun unlockVehicle() {
        val gatt = bluetoothGatt ?: return
        val key = vehicleKey ?: return
        val token = accessToken ?: return

        val timestamp = System.currentTimeMillis()
        val command = """{"action":"unlock","token":"$token","timestamp":$timestamp}"""

        val encryptedData = encryptData(command.toByteArray(), key)

        sendCommand(gatt, encryptedData)

        _unlockStatus.value = UnlockStatus.UNLOCKED
    }

    fun lockVehicle() {
        val gatt = bluetoothGatt ?: return
        val key = vehicleKey ?: return
        val token = accessToken ?: return

        val timestamp = System.currentTimeMillis()
        val command = """{"action":"lock","token":"$token","timestamp":$timestamp}"""

        val encryptedData = encryptData(command.toByteArray(), key)

        sendCommand(gatt, encryptedData)

        _unlockStatus.value = UnlockStatus.LOCKED
    }

    private fun sendCommand(gatt: BluetoothGatt, data: ByteArray) {
        val service = gatt.getService(vehicleServiceUUID)
        val characteristic = service?.getCharacteristic(unlockCharacteristicUUID)

        characteristic?.let {
            it.value = data
            it.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            gatt.writeCharacteristic(it)
            println("Command sent")
        }
    }

    // MARK: - NFC Operations

    fun handleNfcIntent(intent: android.content.Intent) {
        if (!isTokenValid()) {
            println("Access token expired")
            return
        }

        val tag = intent.getParcelableExtra<Tag>(NfcAdapter.EXTRA_TAG) ?: return
        val ndef = Ndef.get(tag)

        ndef?.connect()
        val ndefMessage = ndef?.ndefMessage

        ndefMessage?.records?.forEach { record ->
            val payload = String(record.payload)
            if (payload.startsWith("FLEET:")) {
                val vehicleId = payload.removePrefix("FLEET:")
                println("NFC detected vehicle: $vehicleId")
                unlockViaNFC(vehicleId)
            }
        }

        ndef?.close()
    }

    private fun unlockViaNFC(vehicleId: String) {
        // Simulate NFC unlock
        println("NFC unlock command sent for vehicle $vehicleId")
        _unlockStatus.value = UnlockStatus.UNLOCKED
    }

    // MARK: - Security/Encryption

    private fun encryptData(data: ByteArray, key: SecretKey): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key)

        val iv = cipher.iv
        val encryptedData = cipher.doFinal(data)

        // Combine IV and encrypted data
        return iv + encryptedData
    }

    private fun calculateDistance(rssi: Double): Double {
        val txPower = -59.0 // Calibrated at 1 meter
        if (rssi == 0.0) return -1.0

        val ratio = rssi / txPower
        return if (ratio < 1.0) {
            ratio.pow(10)
        } else {
            0.89976 * ratio.pow(7.7095) + 0.111
        }
    }

    // MARK: - Cloud Backup

    fun requestCloudUnlock(vehicleId: String, callback: (Boolean) -> Unit) {
        // Cloud-based unlock fallback
        // Implementation would use OkHttp or Retrofit
        println("Cloud unlock requested for vehicle $vehicleId")
        callback(true)
    }
}

// MARK: - Data Models

data class VehicleBeacon(
    val id: String,
    val name: String,
    val distance: Double,
    val rssi: Int,
    val device: BluetoothDevice?
)

enum class ConnectionStatus {
    DISCONNECTED,
    CONNECTING,
    CONNECTED
}

enum class UnlockStatus {
    LOCKED,
    UNLOCKING,
    UNLOCKED
}

// MARK: - Jetpack Compose UI

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.LockOpen
import androidx.compose.material3.*
import androidx.compose.ui.graphics.Color

@Composable
fun KeylessEntryScreen(manager: KeylessEntryManager) {
    val isScanning by manager.isScanning.collectAsState()
    val nearbyVehicles by manager.nearbyVehicles.collectAsState()
    val connectionStatus by manager.connectionStatus.collectAsState()
    val unlockStatus by manager.unlockStatus.collectAsState()

    Column(
        modifier = androidx.compose.ui.Modifier
            .fillMaxSize()
            .padding(16.dp),
        horizontalAlignment = androidx.compose.ui.Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        // Connection status
        Row(verticalAlignment = androidx.compose.ui.Alignment.CenterVertically) {
            Box(
                modifier = androidx.compose.ui.Modifier.size(12.dp),
                contentAlignment = androidx.compose.ui.Alignment.Center
            ) {
                Surface(
                    shape = MaterialTheme.shapes.small,
                    color = if (connectionStatus == ConnectionStatus.CONNECTED) Color.Green else Color.Gray
                ) {}
            }
            Spacer(modifier = androidx.compose.ui.Modifier.width(8.dp))
            Text(
                when (connectionStatus) {
                    ConnectionStatus.DISCONNECTED -> "Not Connected"
                    ConnectionStatus.CONNECTING -> "Connecting..."
                    ConnectionStatus.CONNECTED -> "Connected"
                }
            )
        }

        // Unlock button
        Button(
            onClick = {
                if (unlockStatus == UnlockStatus.LOCKED) {
                    manager.unlockVehicle()
                } else {
                    manager.lockVehicle()
                }
            },
            modifier = androidx.compose.ui.Modifier.size(150.dp),
            enabled = connectionStatus == ConnectionStatus.CONNECTED,
            colors = ButtonDefaults.buttonColors(
                containerColor = if (unlockStatus == UnlockStatus.UNLOCKED) Color.Green else Color.Blue
            )
        ) {
            Column(horizontalAlignment = androidx.compose.ui.Alignment.CenterHorizontally) {
                Icon(
                    if (unlockStatus == UnlockStatus.UNLOCKED) Icons.Default.LockOpen else Icons.Default.Lock,
                    contentDescription = "Lock",
                    modifier = androidx.compose.ui.Modifier.size(48.dp)
                )
                Text(if (unlockStatus == UnlockStatus.UNLOCKED) "Lock" else "Unlock")
            }
        }

        // Scan button
        if (isScanning) {
            CircularProgressIndicator()
            Text("Scanning...")
        } else {
            Button(onClick = { manager.startScanning() }) {
                Text("Scan for Vehicles")
            }
        }

        // Nearby vehicles list
        LazyColumn {
            items(nearbyVehicles) { vehicle ->
                Card(
                    modifier = androidx.compose.ui.Modifier
                        .fillMaxWidth()
                        .padding(vertical = 4.dp)
                ) {
                    Row(
                        modifier = androidx.compose.ui.Modifier.padding(16.dp),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = androidx.compose.ui.Alignment.CenterVertically
                    ) {
                        Column {
                            Text(vehicle.name, style = MaterialTheme.typography.bodyLarge)
                            Text(
                                "${"%.1f".format(vehicle.distance)}m away",
                                style = MaterialTheme.typography.bodySmall,
                                color = Color.Gray
                            )
                        }

                        Button(onClick = { manager.connect(vehicle) }) {
                            Text("Connect")
                        }
                    }
                }
            }
        }

        // NFC button
        Button(
            onClick = { /* Trigger NFC */ },
            colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFFF9800))
        ) {
            Text("Unlock via NFC")
        }
    }
}

File: mobile-apps/ios/KeylessEntry.swift
//
//  KeylessEntry.swift
//  Fleet Manager - Keyless Vehicle Entry
//
//  Bluetooth Low Energy and NFC-based keyless entry system
//  with encrypted keys and time-limited access
//

import Foundation
import CoreBluetooth
import CoreNFC
import CryptoKit

// MARK: - Keyless Entry Manager

class KeylessEntryManager: NSObject, ObservableObject {
    static let shared = KeylessEntryManager()

    @Published var isScanning = false
    @Published var nearbyVehicles: [VehicleBeacon] = []
    @Published var connectionStatus: ConnectionStatus = .disconnected
    @Published var unlockStatus: UnlockStatus = .locked

    private var centralManager: CBCentralManager!
    private var nfcSession: NFCNDEFReaderSession?
    private var connectedPeripheral: CBPeripheral?

    // Vehicle service UUID (custom UUID for fleet vehicles)
    private let vehicleServiceUUID = CBUUID(string: "12345678-1234-1234-1234-123456789ABC")
    private let unlockCharacteristicUUID = CBUUID(string: "12345678-1234-1234-1234-123456789ABD")
    private let statusCharacteristicUUID = CBUUID(string: "12345678-1234-1234-1234-123456789ABE")

    // Security
    private var vehicleKey: SymmetricKey?
    private var accessToken: String?
    private var tokenExpiry: Date?

    private override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
        loadStoredCredentials()
    }

    // MARK: - Credential Management

    private func loadStoredCredentials() {
        // Load encrypted vehicle key from keychain
        if let keyData = KeychainHelper.load(key: "vehicle_key") {
            vehicleKey = SymmetricKey(data: keyData)
        }

        // Load access token
        accessToken = UserDefaults.standard.string(forKey: "access_token")

        // Load token expiry
        if let expiryTimestamp = UserDefaults.standard.object(forKey: "token_expiry") as? TimeInterval {
            tokenExpiry = Date(timeIntervalSince1970: expiryTimestamp)
        }
    }

    func setVehicleCredentials(vehicleId: String, key: Data, token: String, expiryHours: Int = 24) {
        // Store encrypted vehicle key
        vehicleKey = SymmetricKey(data: key)
        KeychainHelper.save(key: "vehicle_key", data: key)

        // Store access token
        accessToken = token
        UserDefaults.standard.set(token, forKey: "access_token")

        // Set token expiry
        tokenExpiry = Date().addingTimeInterval(TimeInterval(expiryHours * 3600))
        UserDefaults.standard.set(tokenExpiry?.timeIntervalSince1970, forKey: "token_expiry")

        print("Vehicle credentials set for \(vehicleId), expires at \(tokenExpiry!)")
    }

    private func isTokenValid() -> Bool {
        guard let expiry = tokenExpiry else { return false }
        return Date() < expiry
    }

    // MARK: - Bluetooth Scanning

    func startScanning() {
        guard centralManager.state == .poweredOn else {
            print("Bluetooth not ready")
            return
        }

        guard isTokenValid() else {
            print("Access token expired")
            return
        }

        isScanning = true
        nearbyVehicles.removeAll()

        centralManager.scanForPeripherals(
            withServices: [vehicleServiceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )

        print("Started scanning for vehicles...")
    }

    func stopScanning() {
        centralManager.stopScan()
        isScanning = false
        print("Stopped scanning")
    }

    // MARK: - Vehicle Connection

    func connect(to vehicle: VehicleBeacon) {
        guard let peripheral = vehicle.peripheral else { return }

        connectionStatus = .connecting
        centralManager.connect(peripheral, options: nil)
        print("Connecting to \(vehicle.name)...")
    }

    func disconnect() {
        if let peripheral = connectedPeripheral {
            centralManager.cancelPeripheralConnection(peripheral)
        }
    }

    // MARK: - Unlock/Lock Operations

    func unlockVehicle() {
        guard let peripheral = connectedPeripheral,
              let key = vehicleKey,
              let token = accessToken else {
            print("Not connected or missing credentials")
            return
        }

        // Create unlock command with encrypted payload
        let timestamp = Int(Date().timeIntervalSince1970)
        let command = UnlockCommand(
            action: "unlock",
            token: token,
            timestamp: timestamp
        )

        guard let commandData = try? JSONEncoder().encode(command) else { return }

        // Encrypt command
        guard let encryptedData = try? AES.GCM.seal(commandData, using: key) else {
            print("Encryption failed")
            return
        }

        // Send unlock command
        sendCommand(encryptedData.combined!, to: peripheral)

        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            self.unlockStatus = .unlocked
        }
    }

    func lockVehicle() {
        guard let peripheral = connectedPeripheral,
              let key = vehicleKey,
              let token = accessToken else {
            return
        }

        let timestamp = Int(Date().timeIntervalSince1970)
        let command = UnlockCommand(
            action: "lock",
            token: token,
            timestamp: timestamp
        )

        guard let commandData = try? JSONEncoder().encode(command),
              let encryptedData = try? AES.GCM.seal(commandData, using: key) else {
            return
        }

        sendCommand(encryptedData.combined!, to: peripheral)

        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            self.unlockStatus = .locked
        }
    }

    private func sendCommand(_ data: Data, to peripheral: CBPeripheral) {
        guard let service = peripheral.services?.first(where: { $0.uuid == vehicleServiceUUID }),
              let characteristic = service.characteristics?.first(where: { $0.uuid == unlockCharacteristicUUID }) else {
            print("Service or characteristic not found")
            return
        }

        peripheral.writeValue(data, for: characteristic, type: .withResponse)
        print("Command sent")
    }

    // MARK: - NFC Operations

    func startNFCSession() {
        guard NFCNDEFReaderSession.readingAvailable else {
            print("NFC not available on this device")
            return
        }

        guard isTokenValid() else {
            print("Access token expired")
            return
        }

        nfcSession = NFCNDEFReaderSession(delegate: self, queue: nil, invalidateAfterFirstRead: false)
        nfcSession?.alertMessage = "Hold your iPhone near the vehicle's NFC reader"
        nfcSession?.begin()
        print("NFC session started")
    }

    private func unlockViaNFC(vehicleId: String) {
        guard let key = vehicleKey,
              let token = accessToken else { return }

        // Create unlock command
        let timestamp = Int(Date().timeIntervalSince1970)
        let command = UnlockCommand(
            action: "unlock",
            token: token,
            timestamp: timestamp
        )

        guard let commandData = try? JSONEncoder().encode(command),
              let encryptedData = try? AES.GCM.seal(commandData, using: key) else {
            return
        }

        // In production, this would write to the NFC tag
        // For now, simulate successful unlock
        print("NFC unlock command sent for vehicle \(vehicleId)")

        DispatchQueue.main.async {
            self.unlockStatus = .unlocked
        }
    }

    // MARK: - Auto-lock

    func enableAutoLock(after seconds: Int = 30) {
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(seconds)) { [weak self] in
            guard self?.unlockStatus == .unlocked else { return }
            self?.lockVehicle()
            print("Auto-locked after \(seconds) seconds")
        }
    }

    // MARK: - Cloud Backup

    func requestCloudUnlock(vehicleId: String, completion: @escaping (Bool) -> Void) {
        // Fallback cloud-based unlock if Bluetooth fails
        guard let token = accessToken else {
            completion(false)
            return
        }

        let url = URL(string: "https://api.fleet-manager.com/v1/vehicles/\(vehicleId)/unlock")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let httpResponse = response as? HTTPURLResponse,
               httpResponse.statusCode == 200 {
                DispatchQueue.main.async {
                    self.unlockStatus = .unlocked
                }
                completion(true)
            } else {
                print("Cloud unlock failed: \(error?.localizedDescription ?? "Unknown error")")
                completion(false)
            }
        }.resume()
    }
}

// MARK: - CBCentralManagerDelegate

extension KeylessEntryManager: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            print("Bluetooth ready")
        case .poweredOff:
            print("Bluetooth off")
        case .unsupported:
            print("Bluetooth not supported")
        default:
            break
        }
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        // Parse vehicle info from advertisement data
        let name = advertisementData[CBAdvertisementDataLocalNameKey] as? String ?? "Unknown Vehicle"
        let distance = calculateDistance(rssi: RSSI.doubleValue)

        let vehicle = VehicleBeacon(
            id: peripheral.identifier.uuidString,
            name: name,
            distance: distance,
            rssi: RSSI.intValue,
            peripheral: peripheral
        )

        if !nearbyVehicles.contains(where: { $0.id == vehicle.id }) {
            nearbyVehicles.append(vehicle)
            print("Discovered vehicle: \(name) at \(distance)m")
        }
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        print("Connected to \(peripheral.name ?? "vehicle")")
        connectionStatus = .connected
        connectedPeripheral = peripheral
        peripheral.delegate = self
        peripheral.discoverServices([vehicleServiceUUID])
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        print("Disconnected from vehicle")
        connectionStatus = .disconnected
        connectedPeripheral = nil
    }

    private func calculateDistance(rssi: Double) -> Double {
        // Simple distance estimation based on RSSI
        let txPower = -59.0 // Calibrated at 1 meter
        if rssi == 0 {
            return -1.0
        }

        let ratio = rssi / txPower
        if ratio < 1.0 {
            return pow(ratio, 10)
        } else {
            let distance = (0.89976) * pow(ratio, 7.7095) + 0.111
            return distance
        }
    }
}

// MARK: - CBPeripheralDelegate

extension KeylessEntryManager: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let services = peripheral.services else { return }

        for service in services where service.uuid == vehicleServiceUUID {
            peripheral.discoverCharacteristics(
                [unlockCharacteristicUUID, statusCharacteristicUUID],
                for: service
            )
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard let characteristics = service.characteristics else { return }

        for characteristic in characteristics {
            if characteristic.uuid == statusCharacteristicUUID {
                peripheral.readValue(for: characteristic)
                peripheral.setNotifyValue(true, for: characteristic)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard let data = characteristic.value else { return }

        if characteristic.uuid == statusCharacteristicUUID {
            // Parse vehicle status
            if let status = String(data: data, encoding: .utf8) {
                print("Vehicle status: \(status)")
                DispatchQueue.main.async {
                    self.unlockStatus = status == "unlocked" ? .unlocked : .locked
                }
            }
        }
    }
}

// MARK: - NFCNDEFReaderSessionDelegate

extension KeylessEntryManager: NFCNDEFReaderSessionDelegate {
    func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {
        for message in messages {
            for record in message.records {
                if let payload = String(data: record.payload, encoding: .utf8) {
                    // Parse vehicle ID from NFC tag
                    if payload.hasPrefix("FLEET:") {
                        let vehicleId = payload.replacingOccurrences(of: "FLEET:", with: "")
                        print("NFC detected vehicle: \(vehicleId)")
                        unlockViaNFC(vehicleId: vehicleId)
                    }
                }
            }
        }
    }

    func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        print("NFC session invalidated: \(error.localizedDescription)")
    }
}

// MARK: - Supporting Types

struct VehicleBeacon: Identifiable {
    let id: String
    let name: String
    let distance: Double
    let rssi: Int
    let peripheral: CBPeripheral?
}

enum ConnectionStatus {
    case disconnected
    case connecting
    case connected
}

enum UnlockStatus {
    case locked
    case unlocking
    case unlocked
}

struct UnlockCommand: Codable {
    let action: String
    let token: String
    let timestamp: Int
}

// MARK: - Keychain Helper

class KeychainHelper {
    static func save(key: String, data: Data) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]

        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }

    static func load(key: String) -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true
        ]

        var result: AnyObject?
        SecItemCopyMatching(query as CFDictionary, &result)

        return result as? Data
    }

    static func delete(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key
        ]

        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - UI View

import SwiftUI

struct KeylessEntryView: View {
    @StateObject private var manager = KeylessEntryManager.shared

    var body: some View {
        VStack(spacing: 20) {
            // Connection status
            HStack {
                Circle()
                    .fill(manager.connectionStatus == .connected ? Color.green : Color.gray)
                    .frame(width: 12, height: 12)
                Text(statusText)
                    .font(.subheadline)
            }

            // Unlock button
            Button(action: {
                if manager.unlockStatus == .locked {
                    manager.unlockVehicle()
                } else {
                    manager.lockVehicle()
                }
            }) {
                VStack {
                    Image(systemName: manager.unlockStatus == .unlocked ? "lock.open.fill" : "lock.fill")
                        .font(.system(size: 60))
                    Text(manager.unlockStatus == .unlocked ? "Lock" : "Unlock")
                        .font(.headline)
                }
                .foregroundColor(.white)
                .frame(width: 150, height: 150)
                .background(manager.unlockStatus == .unlocked ? Color.green : Color.blue)
                .clipShape(Circle())
            }
            .disabled(manager.connectionStatus != .connected)

            // Nearby vehicles
            if manager.isScanning {
                ProgressView("Scanning...")
            } else {
                Button("Scan for Vehicles") {
                    manager.startScanning()
                }
            }

            List(manager.nearbyVehicles) { vehicle in
                HStack {
                    VStack(alignment: .leading) {
                        Text(vehicle.name)
                            .font(.headline)
                        Text("\(String(format: "%.1f", vehicle.distance))m away")
                            .font(.caption)
                            .foregroundColor(.gray)
                    }

                    Spacer()

                    Button("Connect") {
                        manager.connect(to: vehicle)
                    }
                }
            }

            // NFC unlock button
            Button("Unlock via NFC") {
                manager.startNFCSession()
            }
            .padding()
            .background(Color.orange)
            .foregroundColor(.white)
            .cornerRadius(10)
        }
        .padding()
    }

    private var statusText: String {
        switch manager.connectionStatus {
        case .disconnected:
            return "Not Connected"
        case .connecting:
            return "Connecting..."
        case .connected:
            return "Connected"
        }
    }
}
